### ISSUE TITLE
Enum properties

### ISSUE THREAD

**Issue by gtmacgregor** [mfidemraizer/trackerdog#19]
Is it possible to get this working with tracked enum properties?

public class A {
 public enum AEnum {
  One = 1,
  Two = 2
 } 

 public virtual string Name { get; set; }
 public virtual AEnum AVal { get; set; }
}

When I try to track the A class (or instance of) it failed with a Contract Exception.

Thanks

 Glenn

**Comment by mfidemraizer** [comment]
@gtmacgregor 

Hi! 

I'll try to reproduce the issue and I'll give you a solution for it as soon as possible.

Thank you for using TrackerDog!

**Comment by mfidemraizer** [comment]
Hi again!

In my case, there was no issue in your scenario:

```
        A a = TrackableObjectFactory.CreateFrom(new A());
        a.AVal = A.AEnum.One;
```

Are you using TrackerDog 2.x? Please verify that you're on 2.x version, because maybe the bug is on 1.x branch, which is obsolete and it won't be more further development on it.

If you're using TrackerDog 2.x, how are you configuring your `A`  class to be trackable? Manual configuration or attribute-based configuration?

And... Will you share the whole `Contract` exception to see the message? Usually the message is enough to understand what's behind the issue.

Thank you in advance! ;)

**Comment by mfidemraizer** [comment]
Since I got no feedback, I close the issue for now. If you want to continue the discussion, feel free to re-open it!

**Comment by gtmacgregor** [comment]
Hi!

Sorry for the long delay. I am using TrackerDog 2.1.1 currently and am still having this same issue, although it might be due to how I am using it. Let me explain my thoughts below:

I have a set of lightweight DTO's which are meant to be readonly interfaces & objects that represent data from a store (DB). The interfaces are readonly meaning the properties only have getters but the implementations have internal setters so the orm can populate as necessary. I don't want to muddle these DTO's with change tracking so I created a set of MutableInterfaces and implementations which just expose setters over the DTO's readonly interface.

```
public interface IUser {
  bool Active {get;}
  string Name {get;}
  UserType UserKind {get;}
}

public class User : IUser {
 ...
 bool Active {
  get {
    return mActive;
  }

  internal set {
    mActive = value;
  }
 }

....
}

IMutableUser : IUser
{
  new string Name {get; set;}
  new bool Active {get; set;}
  new UserType UserKind {get; set;}
}

```

UserType is a public enum.

To construct a MutableUser I pass an IUser in the constructor and deep clone it to populate all the data on the mutableuser.

```
public class MutableUser : User, IMutableUser
{
  public MutableUser(IUser original) : base(original, true)  <-- this is the copy constructor or deep clone constructor
 {
   //setup tracker dog
  ...
  mMutableUser = trackableObjectFactory.CreateFrom(this);
 }
}
 
```
The last line above throws a Contract Exception:

`Assertion failed: (type.IsClass && !type.IsSealed) || type.IsInterface  Given type 'XXX.DataAccess.UserType, ZZZ.DataAccess, Version=1.0.1.0, Culture=neutral, PublicKeyToken=null' must be either a non-sealed class or an interface
`
Not sure if there is a flaw in my logic or a problem with tracking (or not tracking ) enum types.

Thanks

 Glenn

**Comment by mfidemraizer** [comment]
@gtmacgregor Oops. I see. I believe that something like this happened in another issue but we concluded that it was something wrong on the application code. I don't remember the details.

Could you paste the exception *stack trace* too, please? 

Today I won't be able to try your code, but in 1-2 days I'll give you an answer or a bugfix (if it's a bug!).

Thank you for the effort for the detailed explanation!

**Comment by mfidemraizer** [comment]
@gtmacgregor Could you also provide your tracking configuration?

**Comment by gtmacgregor** [comment]
Thanks for the quick response!

So for my testing I am not using attribute based config, below is all I am doing - not I am not even tracking the UserKind property:

this is inside the MutableUser constructor:
```
IObjectChangeTrackingConfiguration config = ObjectChangeTracking.CreateConfiguration();
config.TrackThisType<IMutableAccountExecutive>(t => t.IncludeProperty(a => a.Active));
ITrackableObjectFactory trackableObjectFactory = config.CreateTrackableObjectFactory();
mTestUser = trackableObjectFactory.CreateFrom(this);
```

Below is the stack trace:

```
   at System.Diagnostics.Contracts.__ContractsRuntime.TriggerFailure(ContractFailureKind kind, String msg, String userMessage, String conditionTxt, Exception inner)
   at System.Diagnostics.Contracts.__ContractsRuntime.ReportFailure(ContractFailureKind kind, String msg, String conditionTxt, Exception inner)
   at System.Diagnostics.Contracts.__ContractsRuntime.Assert(Boolean condition, String msg, String conditionTxt)
   at TrackerDog.Configuration.TrackableType..ctor(IObjectChangeTrackingConfiguration configuration, Type type)
   at TrackerDog.Configuration.ObjectChangeTrackingConfiguration.<>c__DisplayClass25_0.<TrackThisTypeRecursive>b__3(Type t)
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at TrackerDog.Configuration.ObjectChangeTrackingConfiguration.TrackThisTypeRecursive(Type rootType, Action`1 configure, TypeSearchSettings searchSettings)
   at TrackerDog.TrackableObjectFactoryInternal.Create(Object some, Type typeToTrack, ObjectChangeTracker reusedTracker, Object parentObject, PropertyInfo propertyToSet, Object[] constructorArguments)
   at TrackerDog.TrackableObjectFactoryInternal.Create[TObject](TObject some, ObjectChangeTracker reusedTracker, Object parentObject, PropertyInfo propertyToSet, Object[] constructorArguments)
   at TrackerDog.TrackableObjectFactoryInternal.CreateFrom[TObject](TObject some)
   at PHC.BatchSource.DataAccess.Model.MutableAccountExecutive..ctor(IAccountExecutive original) in C:\PHC\projects-git\BatchSource\PHC.BatchSource\PHC.BatchSource.DataAccess\Model\AccountExecutive.cs:line 252
   at PHC.BatchSource.UnitTest.DataAccess.ModelTests.TestMutableAccountExecutiveSettors() in C:\PHC\projects-git\BatchSource\PHC.BatchSource\PHC.BatchSource.UnitTest\DataAccess\ModelTests.cs:line 126
```

**Comment by mfidemraizer** [comment]
@gtmacgregor 

No problem!

Well I'll investigate what's going on in 1 or 2 days

**Comment by mfidemraizer** [comment]
@gtmacgregor  

I've tried to reproduce your scenario... BTW, I'm unable to do so because you didn't provide me the full code, it's like a pseudo-code...

It would be nice if you could attach a ZIP which I could execute in my machine... (for example, you could try to reproduce your code in a sample test project to let me get the contract assertion fail and check what's going on).

**Comment by mfidemraizer** [comment]
@gtmacgregor Nevermind, I could reproduce the issue adding an `enum` property in an interface tracking test. Now I can solve the issue ;P

**Comment by mfidemraizer** [comment]
@gtmacgregor  Good news! I've already fixed it. When configuring change trackings using interfaces, there was an incomplete `Type` filter which wasn't throwing out `enum` (and also `struct`).

I've already published a new NuGet package version: 2.1.2.

Try it out and let me know what happened.

BTW, here I put my two cents about your code:

1. Change tracking configuration shouldn't be created and configured in a class constructor. It should be placed somewhere during your app/system/service initialization, and you should create a trackable object factory to inject wherever you need to turn your configured types into change-trackable objects.

2. Remember that classes should be open (non-sealed) and properties must be `virtual` to be able to track them with TrackerDog.

I'll look forwards for your comments about if the new TrackerDog version has solved your issue, or if you've found new ones (which would need to be added as new issues!).

**Comment by gtmacgregor** [comment]
Thank you, it works as expected now!

A bit of explanation on why I set it up as I did.

I want to have a lightweight readonly version of the DTO which does not rely on external dependencies (TrackerDog) so to do this I have an interface which is readonly (no setters). The implementation of that interface contains internal setters for each property. Then I created a mutable implementation which derives from the readonly version exposing those internal setters as public. If I only had one implementation which was read write then I could not force the end users of the library to get a tracked version to make changes. 

I would be interested in hearing if you have any other solutions to this issue.

Thanks

 Glenn
